UriGenerator.php

```diff
@@ -64,19 +65,35 @@ public function generateUri(UriContext $uriContext)
                     $name, $metadata->getUriSchema()
                 ));
             }
-            $options = $tokenProviderConfigs[$name];
+            $tokenProviderConfig = $tokenProviderConfigs[$name];
 
-            $tokenProvider = $this->serviceRegistry->getTokenProvider($options['name']);
+            $tokenProvider = $this->serviceRegistry->getTokenProvider($tokenProviderConfig['name']);
 
-            // I can see the utility of making this a singleton, but it is a massive
-            // code smell to have this in a base class and be also part of the interface
             $optionsResolver = new OptionsResolver();
+            $this->configureGlobalOptions($optionsResolver);
             $tokenProvider->configureOptions($optionsResolver);
+            $tokenProviderOptions = $tokenProviderConfig['options'];
+
+            $tokenValue = $tokenProvider->provideValue($uriContext, $optionsResolver->resolve($tokenProviderOptions));
+
+            $isEmpty = empty($tokenValue) || $tokenValue == '/';
+
+            if ($isEmpty && false === $tokenProviderOptions['collapse_empty']) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Token provider "%s" returned an empty value for token "%s" with URI schema "%s"',
+                    $tokenProviderConfig['name'], $name, $uriSchema
+                ));
+            }
+
+            $tokenString = '{' . $name . '}';
 
-            $tokens['{' . $name . '}'] = $tokenProvider->provideValue($uriContext, $optionsResolver->resolve($options['options']));
+            if ($isEmpty && true === $tokenProviderOptions['collapse_empty']) {
+                $tokens[$tokenString . '/'] = (string) $tokenValue;
```

Comment:
We treat `/` as a de facto empty value and apply the same logic as we would for an empty string.